### DISCUSSION TITLE
Access peripherals from different modules

### DISCUSSION THREAD

**Discussion by keymusic** [Rahix/avr-hal#609] [Q&A] [answered]
Hi Rust experts and AVR fans,

First of all: Thank you for providing this Rust abstraction layer for the AVR controllers. The examples are very helpful to get started.

For some years I have been using these 8-bit controllers in several applications. I wrote code in Assembler, C, C++ and with the Arduino framework.
The last months I spent some time to learn Rust. I am particularly interested in embedded applications.
I could successfully implement a small application on an Arduino with LEDs, push-buttons and some serial output.
Now, I try to refactor the application and move code to dedicated modules:
- one module for LEDs
- one module for push-buttons
- one module for serial output

The main function should initialize these modules.

I did not manage it to to use different peripherals in different modules.
If I understand it correctly, only one function can take all the peripherals.

```
let dp = arduino_hal::Peripherals::take().unwrap();
let pins = arduino_hal::pins!(dp);
```

So, for the serial peripheral I did it similarly as described in https://github.com/Rahix/avr-hal/issues/115 where Rahix proposed a solution on 2022-01-28.

In the main code I pass the serial peripheral to my console module:
```
let serial = arduino_hal::default_serial!(dp, pins, 57600);
console::console::init(serial);
```
But I get the following messages

> warning: Linking globals named 'DEVICE_PERIPHERALS': symbol multiply defined!
>
> error: failed to load bitcode of module "avr_device-f61410d5a9432922.avr_device.2c1c7d05340c00bc-cgu.0.rcgu.o": 

Main questions:
1. How can I fix this error?
2. Is it possible at all to move the ownership for certain peripherals to modules in other files?

Further questions:
3. What is the granularity level for peripherals? Can I move the ownership of a certain port or pin to another function or module?
4. Do you have further example code where peripherals are accessed from different modules?

Thanks for any hints.
michaeL

**Comment by stappersg** [comment]
> How can I fix this error?

:smile: _plus_ https://duckduckgo.com/?t=ftsa&q=rust++error%3A+failed+to+load+bitcode+of+module&ia=web


Another approach could be providing a _"git clone URL"_ to make reproducing the error possible.

**Comment by keymusic** [comment]
Thank you, @stappersg, for your prompt suggestions. - Sorry, for my delayed response. 

I wrote a simple application which demonstrates my problem. It also appears if the macro for serial console output is implemented in the file where it is used.

Please have a look at [https://github.com/keymusic/advent_wreath/blob/linker_error/src/main.rs](https://github.com/keymusic/advent_wreath/blob/linker_error/src/main.rs).

When I call `cargo build` I get the following messages:

> warning: Linking globals named 'DEVICE_PERIPHERALS': symbol multiply defined!
> error: failed to load bitcode of module "avr_device-f61410d5a9432922.avr_device.2c1c7d05340c00bc-cgu.0.rcgu.o": 
> warning: `advent_wreath` (bin "advent_wreath") generated 1 warning

**Reply by Rahix** [reply, marked as answer]
Hi,

I think you are running into a very nasty cargo behavior that isn't obvious at all:  Looking at your dependencies [here](https://github.com/keymusic/advent_wreath/blob/linker_error/Cargo.toml#L19-L26):

```toml
[dependencies.avr-device]
version = "0.6.0"
features = ["atmega328p"]

[dependencies.arduino-hal]
git = "https://github.com/rahix/avr-hal"
rev = "190f2c3cb8d29e10f71119352b912369dc5a1fb7"
features = ["arduino-uno"]
```

While you are depending on `avr-device=0.6`, the version of `arduino-hal` you are pulling in specifies this:

https://github.com/Rahix/avr-hal/blob/190f2c3cb8d29e10f71119352b912369dc5a1fb7/arduino-hal/Cargo.toml#L54-L55

Cargo now adds **both** versions of `avr-device` to your dependency tree to satisfy these requirements.  But because `avr-device` is a low-level crate, it defines things like hardware singletons, as expressed by the `DEVICE_PERIPHERALS` symbol.  Because two versions of `avr-device` try to define this symbol, you get the error you are observing.

The fix is simple, just depend on `avr-device=0.5.1` in your crate as well.

**Reply by keymusic** [reply]
Hi Rahix,

Thank you very much for your support. Yes, this was the reason. The fix was as simple as you have described.
I think, it would have tanken me a long time to find the cause, if I had found the it at all.
With your explanations, it is totally plausible, why the linker error occurs.

**Comment by Rahix** [comment]
Regarding your other questions:

> Is it possible at all to move the ownership for certain peripherals to modules in other files?

The common pattern is to initialize peripherals in `main()` and then move them into the respective module struct.

> What is the granularity level for peripherals? Can I move the ownership of a certain port or pin to another function or module?

After initializing the `Pins` struct with the `pins!()` macro, you can move individual pins to whereever you like.

> Do you have further example code where peripherals are accessed from different modules?

Not really, I am afraid.  But it shouldn't be too difficult to figure out: Just create structs for each "module" and move the required peripherals into those structs.

**Reply by keymusic** [reply]
Also thanks for these conceptual explanations.
Maybe I find some time after the Christmas Holidays ("between the years") to play around with this.
Please have a good festive season!
